gkctl/main.c

```diff
@@ -0,0 +1,265 @@
+
+static struct argp_option options[] = {
+	{"server-path",		's',	"FILE",		0,
+		"Path to Gatekeeper/Grantor's UNIX socket",	1},
+	{ 0 }
+};
+
+struct args {
+	const char *server_path;
+	const char *lua_script_path;
+};
+
+static error_t
+parse_opt(int key, char *arg, struct argp_state *state)
+{
+	struct args *args = state->input;
+
+	switch (key) {
+	case 's':
+		args->server_path = arg;
+		break;
+
+	case ARGP_KEY_INIT:
+		args->lua_script_path = NULL;
+		break;
+
+	case ARGP_KEY_ARG:
+		if (args->lua_script_path) {
+			argp_error(state,
+				"Wrong number of arguments; only one is allowed");
+		}
+		args->lua_script_path = arg;
+		break;
+
+	case ARGP_KEY_END:
+		if (!args->lua_script_path) {
+			argp_error(state,
+				"The lua script path was not specified");
+		}
+		break;
+
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+
+	return 0;
+}
+
+static struct argp argp = {options, parse_opt, adoc, doc, NULL, NULL, NULL};
+
+static int
+load_file_to_buffer(const char *file_name, char *buffer, int n)
+{
+	int ret;
+	FILE *fp = fopen(file_name, "r");
+	if (fp == NULL) {
+		fprintf(stderr, "Error: Failed to open file %s - %s\n",
+			file_name, strerror(errno));
+		ret = -1;
+		goto out;
+	}
+
+	/*
+	 * Return value equals the number of bytes transferred
+	 * only when size (i.e., second parameter) is 1.
+	 */
+	ret = fread(buffer, 1, n, fp);
+	if (ferror(fp)) {
+		fprintf(stderr, "Error: %s\n", strerror(errno));
+		ret = -1;
+	} else if (ret == n && !feof(fp)) {
+		fprintf(stderr, "Error: Failed to read the whole file %s (file length exceeds the maximum message size - %d)\n",
+			file_name, n);
+		ret = -1;
+	}
+
+	fclose(fp);
+out:
+	return ret;
+}
+
+static int
+write_all(int conn_fd, const char *msg_buff, int nbytes)
+{
+	int send_size;
+	int tot_size = 0;
+
+	if (nbytes <= 0)
+		return 0;
+
+	while ((send_size = write(conn_fd, msg_buff + tot_size,
+			nbytes - tot_size)) > 0) {
+		tot_size += send_size;
+		if (tot_size >= nbytes)
+			break;
+	}
+
+	/* The connection with the server is closed. */
+	if (send_size == 0) {
+		fprintf(stderr, "Server disconnected\n");
+		return -1;
+	}
+
+	if (send_size < 0) {
+		fprintf(stderr, "Failed to write data to the socket connection - (%s)\n",
+			strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+static int
+read_all(int conn_fd, char *msg_buff, int nbytes)
+{
+	int recv_size;
+	int tot_size = 0;
+
+	if (nbytes <= 0)
+		return 0;
+
+	while ((recv_size = read(conn_fd, msg_buff + tot_size,
+			nbytes - tot_size)) > 0) {
+		tot_size += recv_size;
+		if (tot_size >= nbytes)
+			break;
+	}
+
+	/* The connection with the server is closed. */
+	if (recv_size == 0) {
+		fprintf(stderr, "Server disconnected\n");
+		return -1;
+	}
+
+	if (recv_size < 0) {
+		fprintf(stderr, "Failed to read data from the socket connection - (%s)\n",
+			strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+int
+main(int argc, char *argv[])
+{
+	int ret;
+	int sock_fd;
+	char send_buff[MSG_MAX_LEN + 2];
```

Comment:
Replace `2` with `sizeof(uint16_t)` since we're already using `sizeof(uint16_t)` everywhere else.